pennylane/operation.py

```diff
@@ -906,6 +906,50 @@ def __matmul__(self, other):
 
         raise ValueError("Can only perform tensor products between observables.")
 
+    def obs_data(self):
+        r"""Extracts the data from an Observable/Tensor."""
+        obs = Tensor(self).non_identity_obs
+        tensor = set()
+
+        for ob in obs:
+            parameters = tuple(param.tostring() for param in ob.parameters)
+            tensor.add((ob.name, ob.wires, parameters))
+
+        return tensor
+
+    def compare(self, other):
```

Comment:
Worth making this `__eq__`?

Reply:
A side effect will be that Hamiltonian objects are no longer hashable (https://hynek.me/articles/hashes-and-equality/) which _might_ be a good thing

Reply:
We originally implemented it like this for Hamiltonians and Observables. However, when we tried to run the tests, we found that overloading equality for the Observable class broke a bunch of stuff, so we decided to change it to `compare()`. We also went from `__eq__` to `compare()` in the Hamiltonian class for the sake of consistency.

Reply:
Ah, this is probably a side effect of Hamiltonians not being hashable, making them unable to be used in tuples, set, etc.

Reply:
Hm, possibly, although I think it was more of an issue with the Observable class. There seem to be instances in the code that rely upon the fact that each created Observable is unique (even if all of its parameters are equal to the parameters of another observable), and when we overloaded equality, we broke this!